File: base/components/node.cc
#include "base/components/node.h"

Node::Node() : erase_mark(false) {}
Node::~Node() { Clear(); }

void Node::Update() {}
void Node::Render() {}

void Node::UpdateChildren() {
  Update();
  for (auto it = children.begin(); it != children.end(); it++) {
    (*it)->UpdateChildren();
  }
}

void Node::RenderChildren() {
  Render();
  for (auto it = children.begin(); it != children.end(); it++) {
    (*it)->RenderChildren();
  }
}

void Node::RemoveMarkedChild() {
  for (auto it = children.begin(); it != children.end();) {
    if ((*it)->IsErase()) {
      it = children.erase(it);
    } else {
      it++;
    }
  }

  for (auto it = children.begin(); it != children.end(); it++) {
    (*it)->RemoveMarkedChild();
  }
}

void Node::AddChild(std::shared_ptr<Node> child) {
  child->parent = this;
  children.push_back(child);
}

void Node::Clear() { children.clear(); }

File: builder/node.py
class Node:
    def __init__(self, name):
        self.children = []
        self.parent = []
        self.name = name

    @property
    def isroot(self):
        return self.parent.count() == 0

    def add_child(self, node):
        node.__add_parent(self)
        self.children.append(node)

    def __add_parent(self, node):
        self.parent.append(node)

    def __str__(self):
        return self.name
